[Enhancement] Child Link Can Hold Hylian Shield

Patrick12115 opened the pull request:
Adds an enhancement to hold the Hylian shield as Child Link, the same way Adult Link does. Has an additional option to rotate and scale the shield to sit on Link's back as the rest of the shields do. This uses a patching method which will make mods that replace the shield show Master Sword instead of Kokiri sword when this is enabled, so mods are recommended to keep this setting off.

Separated from #3830 to be its own PR

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2514151118.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2514155367.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2514165101.zip)
<!--- section:artifacts:end -->

Patrick12115 committed:
Moved to own branch

Patrick12115 committed:
Moved to own branch plus some tweaks

inspectredc commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -1337,6 +1339,22 @@ void DrawCheatsMenu() {
         UIWidgets::PaddedEnhancementCheckbox("Shield with Two-Handed Weapons", "gShieldTwoHanded", true, false);
         UIWidgets::Tooltip("This allows you to put up your shield with any two-handed weapon in hand except for Deku Sticks");
+        if (UIWidgets::PaddedEnhancementCheckbox("Hold Hylian Shield as Child Link", "gChildHoldsHylianShield", true, false)) {
+            if (CVarGetInteger("gChildHoldsHylianShield", 0) == 0) {
+            	CVarSetInteger("gRotateScaleChildHylianShield", 0);
+            }
+            auto player = GET_PLAYER(gPlayState);
+            Player_SetModels(player, Player_ActionToModelGroup(player, player->heldItemAction));
+        }
+        UIWidgets::Tooltip("Allows Child Link to hold the Hylian Shield the same way as the rest of the shields.");
+        if (CVarGetInteger("gChildHoldsHylianShield", 0) == 1) {
+			if (UIWidgets::PaddedEnhancementCheckbox("Rotate and Scale Child Hylian Shield", "gRotateScaleChildHylianShield", true, false)) {
+                auto player = GET_PLAYER(gPlayState);
+                Player_SetModels(player, Player_ActionToModelGroup(player, player->heldItemAction));
+                UpdatePatchChildHylianShield();
+            }
+		}
```
Could this be moved to a separate function, maybe in mods.cpp?

Patrick12115 replied:
Not that I know of? But I don't know a lot. The point is to have it run the functions to draw properly when the checkbox is changed. So if there is a better way, I don't know it haha

inspectredc replied:
Yea i just meant like, in SohMenuBar, call smth like 'UpdateChildHylianShieldState' which is a function defined in mods.cpp which runs this code here

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -580,7 +580,11 @@ s32 func_8008E9C4(Player* this) {
-    return gSaveContext.linkAge != 0 && (this->currentShield == PLAYER_SHIELD_HYLIAN);
+    if (CVarGetInteger("gChildHoldsHylianShield", 0)) {
+        return 0;
+    } else {
```
Perhaps a comment here explaining that this is the function which is checked against when deciding to run vanilla child hylian shield logic. Also returning false would be better than 0

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -600,7 +604,11 @@ void Player_SetModelsForHoldingShield(Player* this) {
             this->rightHandType = PLAYER_MODELTYPE_RH_SHIELD;
-            this->rightHandDLists = &sPlayerDListGroups[PLAYER_MODELTYPE_RH_SHIELD][gSaveContext.linkAge];
+            if (LINK_IS_CHILD && (CVarGetInteger("gChildHoldsHylianShield", 0)) && (this->currentShield == PLAYER_SHIELD_HYLIAN)) {
+                    this->rightHandDLists = &sPlayerDListGroups[PLAYER_MODELTYPE_RH_SHIELD][0];
```
```suggestion
                    this->rightHandDLists = &sPlayerDListGroups[PLAYER_MODELTYPE_RH_SHIELD][LINK_AGE_ADULT];
```

I think this would be better, and perhaps a comment too saying we want to treat this like adult hylian shield (assuming I'm right in what's happening here)

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -622,6 +630,10 @@ void Player_SetModels(Player* this, s32 modelGroup) {
     this->rightHandType = gPlayerModelTypes[modelGroup][PLAYER_MODELGROUPENTRY_RIGHT_HAND];
     this->rightHandDLists = &sPlayerDListGroups[this->rightHandType][gSaveContext.linkAge];
 
+    if (LINK_IS_CHILD && CVarGetInteger("gChildHoldsHylianShield", 0) &&
+        this->rightHandType == PLAYER_MODELTYPE_RH_SHIELD && this->currentShield == PLAYER_SHIELD_HYLIAN) {
+        this->rightHandDLists = &sPlayerDListGroups[this->rightHandType][0];
+    }
```
```suggestion
    if (LINK_IS_CHILD && CVarGetInteger("gChildHoldsHylianShield", 0) &&
        this->rightHandType == PLAYER_MODELTYPE_RH_SHIELD && this->currentShield == PLAYER_SHIELD_HYLIAN) {
        this->rightHandDLists = &sPlayerDListGroups[this->rightHandType][LINK_AGE_ADULT];
    }
```

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -630,6 +642,13 @@ void Player_SetModels(Player* this, s32 modelGroup) {
+    if (CVarGetInteger("gRotateScaleChildHylianShield", 0)) {
+        if (LINK_IS_CHILD && this->sheathType == PLAYER_MODELTYPE_SHEATH_18 && this->currentShield == PLAYER_SHIELD_HYLIAN &&
+            ((gSaveContext.equips.buttonItems[0] == ITEM_SWORD_KOKIRI || gSaveContext.equips.buttonItems[0] == ITEM_NONE || gSaveContext.equips.buttonItems[0] == ITEM_STICK))) {
+            this->sheathDLists = &sPlayerDListGroups[this->sheathType][0];
```
```suggestion
            this->sheathDLists = &sPlayerDListGroups[this->sheathType][LINK_AGE_ADULT];
```

inspectredc commented on soh/src/code/z_player_lib.c:
```diff
@@ -1188,6 +1207,27 @@ void func_8008F87C(PlayState* play, Player* this, SkelAnime* skelAnime, Vec3f* p
 s32 Player_OverrideLimbDrawGameplayCommon(PlayState* play, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, void* thisx) {
     Player* this = (Player*)thisx;
 
+    if (CVarGetInteger("gChildHoldsHylianShield", 0) && CVarGetInteger("gRotateScaleChildHylianShield", 0) &&
+        LINK_IS_CHILD) {
+        if (limbIndex == PLAYER_LIMB_SHEATH) {
+            if ((this->currentShield == PLAYER_SHIELD_HYLIAN &&
+                 (gSaveContext.equips.buttonItems[0] == ITEM_SWORD_KOKIRI ||
+                  gSaveContext.equips.buttonItems[0] == ITEM_NONE ||
+                  gSaveContext.equips.buttonItems[0] == ITEM_STICK)) &&
+                ((this->sheathType == PLAYER_MODELTYPE_SHEATH_18) ||
+                 (this->sheathType == PLAYER_MODELTYPE_SHEATH_19))) {
+                Matrix_Translate(218, 0, 62, MTXMODE_APPLY);
+                Matrix_Scale(0.8, 0.8, 0.8, MTXMODE_APPLY);
+            }
+        }
+        if (limbIndex == PLAYER_LIMB_R_HAND) {
+            if (this->currentShield == PLAYER_SHIELD_HYLIAN && sRightHandType == PLAYER_MODELTYPE_RH_SHIELD &&
+                gSaveContext.equips.buttonItems[0] == ITEM_SWORD_KOKIRI) {
+                Matrix_Scale(0.8, 0.8, 0.8, MTXMODE_APPLY);
+            }
+        }
+    }
```
Maybe not for this PR but maybe we could add a hook here for player override limb draw common so that future modifications like this dont stack up in this function

inspectredc commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -687,7 +689,29 @@ void RegisterMirrorModeHandler() {
+        UpdatePatchChildHylianShield();
+    });
+}
+        f32 triforcePieceScale;
```
some odd padding here btw

inspectredc commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -1337,6 +1339,22 @@ void DrawCheatsMenu() {
         UIWidgets::Tooltip("Prevents the Deku Shield from burning on contact with fire");
         UIWidgets::PaddedEnhancementCheckbox("Shield with Two-Handed Weapons", "gShieldTwoHanded", true, false);
         UIWidgets::Tooltip("This allows you to put up your shield with any two-handed weapon in hand except for Deku Sticks");
+        if (UIWidgets::PaddedEnhancementCheckbox("Hold Hylian Shield as Child Link", "gChildHoldsHylianShield", true, false)) {
```
I'd say this belongs in the enhancements->gameplay->items menu more than the cheats menu

Patrick12115 replied:
I put it in cheats since it can be used to break logic, but I'm not opposed to moving it

inspectredc reviewed (changes requested):
Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly

Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.

Patrick12115 commented:
> Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly
> 
> Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.

Just tried downloading it and it doesn't crash for me. What system are you using? Also not sure what you mean by the cheat menu being distorted.

inspectredc commented:
> > Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly
> > Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.
> 
> Just tried downloading it and it doesn't crash for me. What system are you using? Also not sure what you mean by the cheat menu being distorted.

This is on mac, Maybe we just have so many cheats now that the menu wraps around on my laptop for a second column but that's what im getting haha

Patrick12115 commented:
> > > Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly
> > > Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.
> > 
> > 
> > Just tried downloading it and it doesn't crash for me. What system are you using? Also not sure what you mean by the cheat menu being distorted.
> 
> This is on mac, Maybe we just have so many cheats now that the menu wraps around on my laptop for a second column but that's what im getting haha

The two columns is how it is now with changes from #3481 , unless something else weird is going on for you? Not sure without seeing what you're seeing.

inspectredc commented:
> > > > Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly
> > > > Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.
> > > 
> > > 
> > > Just tried downloading it and it doesn't crash for me. What system are you using? Also not sure what you mean by the cheat menu being distorted.
> > 
> > 
> > This is on mac, Maybe we just have so many cheats now that the menu wraps around on my laptop for a second column but that's what im getting haha
> 
> The two columns is how it is now with changes from #3481 , unless something else weird is going on for you? Not sure without seeing what you're seeing.

Ahh ok I just haven't been on develop yet, I think I thought it was related to the crash at first

Patrick12115 commented:
> > > > > Not sure what's going wrong but this crashes for me when i try and select the enhancement. The checkbox also seems to distort the cheats menu somewhat? I've left some general other notes here too, but it generally looks fine, just can't test because of the crash sadly
> > > > > Also, totally sorry for complaining about this not being separated when I didn't know it hadnt been haha.
> > > > 
> > > > 
> > > > Just tried downloading it and it doesn't crash for me. What system are you using? Also not sure what you mean by the cheat menu being distorted.
> > > 
> > > 
> > > This is on mac, Maybe we just have so many cheats now that the menu wraps around on my laptop for a second column but that's what im getting haha
> > 
> > 
> > The two columns is how it is now with changes from #3481 , unless something else weird is going on for you? Not sure without seeing what you're seeing.
> 
> Ahh ok I just haven't been on develop yet, I think I thought it was related to the crash at first

Still no clue about the crash though. I'm on windows, so maybe it's something mac related, but I have no way of testing that

inspectredc commented:
Yea ill pull the branch and check tomorrow

Patrick12115 committed:
Various changes

Moved to enhancements menu, under gameplay and items. Added LINK_IS_ADULT. Added update to use in sohmenubar.

Malkierian commented:
Assuming these are listed in the cheats, since you put [Cheat] in the PR title, I would like to request that you rename the CVars you chose to use to have `gCheats.` as a prefix, ala `gCheats.ChildHoldsHylianShield`.

Patrick12115 commented:
> Assuming these are listed in the cheats, since you put [Cheat] in the PR title, I would like to request that you rename the CVars you chose to use to have `gCheats.` as a prefix, ala `gCheats.ChildHoldsHylianShield`.

I am going to move it to gameplay items, but I assume that means gEnhancements instead, so I will do

Patrick12115 committed:
gEnhancements and other tweaks

Patrick12115 committed:
More tweaks

Patrick12115 committed:
Merge remote-tracking branch 'upstream/develop' into Child-Holds-Hylian-Shield

Patrick12115 committed:
Fixed Crash

garrettjoecox commented on soh/soh/Enhancements/mods.cpp:
```diff
@@ -689,6 +689,43 @@ void RegisterMirrorModeHandler() {
+}
+
+void RegisterPatchChildHylianShieldHandler() {
+    GameInteractor::Instance->RegisterGameHook<GameInteractor::OnPlayerUpdate>([]() {
```
We probably don't want to be patching this every player update

Patrick12115 committed:
Update presets.h

Patrick12115 committed:
Merge remote-tracking branch 'upstream/develop' into Child-Holds-Hylian-Shield

Patrick12115 committed:
Changed update condition

Patrick12115 committed:
Merge remote-tracking branch 'upstream/develop' into Child-Holds-Hylian-Shield

Patrick12115 committed:
Update presets.h

Patrick12115 committed:
CVAR_ENHANCEMENT

Patrick12115 committed:
More

Patrick12115 committed:
Fixes

Pepper0ni commented:
As someone not well versed with the modding scene, why would this be preferred to having a smaller version of the shield over the deku shield and then using the existing fireproof deku shield cheat? In theory that has the same mechanical effect but doesn't need an extra checkbox or kill the ability to use the larger shield for death mountain trail.

Caladius commented:
Alright let me explain this to you:

Imagine you just got to Wendy's and you order a 12piece nugget meal. However due to no real reason at all you can only eat 6 now and the other 6 in 7 years. It's kind of like that.

This grants the ability to modify both shields that Child Link can equip and also be able to utilize both. Not implementing this for the sake of using the Fireproof Shield and simply modding the Deku Shield eliminates an entire shield that Child Link could use. I am also not a modder but I feel like "It doesnt need an extra checkbox" is hypocritical considering some of the things that already exist in the Menubar that add cool things but not necessarily "needed".

Not arguing, just providing feedback <3

Pepper0ni commented:
While I certainly understand the logic of "we already have so many settings what's 1 more", I think that when reviewing we should be asking that kind of question as there is a softcap for how many micro-options we can have in screen real estate and player patience. When I ask about this I'm not demanding a huge difference, we have tons of minimal tickboxes already, just a small use-case to confirm it's not frivolous.

I think that a mod asking players to remove the (admitingly tiny) use-case of the child hylian shield for making what is effectivly a second fireproof deku shield is strange, and want to make sure it's something that will actually come up instead of being a modder crutch that limits players.

Caladius commented:
Agreed 100%, honestly I believe enhancing the modding experience is something that would be beneficial, whether its by addition of this PR or by means of improving the capabilities that SoH provides as a whole. 

Id imagine this PR would be the option most would agree on since the alternative is something that hasn't seen much progress especially with all of the other In-Progress work going on currently.

briaguya0 commented:
i guess i just don't fully understand what this thing does, it seems we have

| shield | how link holds it | can it burn |
| - | - | - |
| deku | in hands | yes (unless using fireproof cheat) |
| hylian | like a turtle | no |

and this changes it to

| shield | how link holds it | can it burn |
| - | - | - |
| deku | in hands | yes (unless using fireproof cheat) |
| hylian | like a turtle (unless using can hold cheat) | no |

it seems to me we could simplify stuff by having

| shield | how link holds it | can it burn |
| - | - | - |
| deku | toggle: in hands / like a turtle | toggle: yes / no |
| hylian | toggle: in hands / like a turtle | toggle: yes / no |

Patrick12115 commented:
It's meant to be an in game upgrade to the deku shield. Instead of using cheats. It's also meant to go along with the Child Link using Adult Link items feature that was already merged. I just separated the PR's for easier reviews. This has also been a requested feature many times. I'm also pretty sure you can still block the falling rocks on DMT, though it has been a very long time since I tested all this. And I mean, sure we could have deku shield do the turtle thing, but that seems counter productive to the issue of saying the option isn't even needed as that would just be adding an extra toggle to a feature I very much doubt would ever get used. It also opens the door for more flexibility in mods, as you can still have say the deku shield and then the heros shield as an upgrade.

Jepvid commented:
Imma throw in my 2 cents in here. I see this toggle as a good thing compared to fireproof dekushield.

It adds a tiny amount of value to the Hylianshield as child link as it makes it useable in deku tree instead of having to find the dekushield. This PR adds some more accessibility in that sense. Making both shields viable for childlink in a randomizer is better than having turtle link.

Pepper0ni commented:
If it is well requested that is enough justification for me, and having it as a different way to handle child shield upgrades in rando when we get to that point is a nice bonus.

briaguya0 commented:
to be clear i'm not opposed to the functionality. i was just thinking of it from a "what properties of shields can be modified" perspective, and proposing we make the same properties of all shields modifiable instead of one property on deku shield and a different property on hylian shield.

Pepper0ni commented:
Hylian shield does not reflect deku nuts from scrubs while held
The description of the hold shield enhancement is that of the rotate and scale option until it is enabled.
The option is listed among the item enhancements when the PR implies it should be a cheat.
When I start the game for the first time, load an empty save with the rotate and scale option on, equip hylian shield, then equip kokiri sword, I get a wierd broken Master sword graphic only visible from behind instead of a Ksowrd graphic. Repeat attempts usually give no graphic at all or the intended graphic. 
![image](https://github.com/HarbourMasters/Shipwright/assets/93387759/5d70bc7d-4e16-4aec-9fe2-bb419ad06970)

> I'm also pretty sure you can still block the falling rocks on DMT, though it has been a very long time since I tested all this

You can get a block to trigger but the breaking hitbox/a second boulder hits you anyway even if you hold the shield upwards, so it does not avoid damage the same way turtle shield would

Patrick12115 committed:
Merge branch 'develop' into Child-Holds-Hylian-Shield

Extloga commented on soh/soh/SohMenuBar.cpp:
```diff
@@ -714,6 +714,20 @@ void DrawEnhancementsMenu() {
+                        UpdatePatchChildHylianShield();
+                    }
+                }
+                UIWidgets::Tooltip("Rotates and scales the hylian shield for Child Link, so that it is the same orientation as the other shields. May not work properly with some mods.");
```
Typo: `hylian shield` &#8594; `Hylian Shield`

Patrick12115 committed:
Merge branch 'develop' into Child-Holds-Hylian-Shield

Patrick12115 committed:
Various fixes

Fixes tool tip, fixed handle patching issue, added deku nut reflect, fixed scaling issues when using MS or BGS

Patrick12115 commented:
> Hylian shield does not reflect deku nuts from scrubs while held The description of the hold shield enhancement is that of the rotate and scale option until it is enabled. The option is listed among the item enhancements when the PR implies it should be a cheat. When I start the game for the first time, load an empty save with the rotate and scale option on, equip hylian shield, then equip kokiri sword, I get a wierd broken Master sword graphic only visible from behind instead of a Ksowrd graphic. Repeat attempts usually give no graphic at all or the intended graphic. ![image](https://private-user-images.githubusercontent.com/93387759/331882676-5d70bc7d-4e16-4aec-9fe2-bb419ad06970.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzgyNzcwMDMsIm5iZiI6MTczODI3NjcwMywicGF0aCI6Ii85MzM4Nzc1OS8zMzE4ODI2NzYtNWQ3MGJjN2QtNGUxNi00YWVjLTlmZTItYmI0MTlhZDA2OTcwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTMwVDIyMzgyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyZDAyODA3OWZiZjU5NjE4YjQzNGY2NWJkNTFlMWE4YTM3N2U2YjkzYmMxNDZjYzI1YzNhMDY3MGRhNmViMzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.e4FOL_JmYjW6gJNT9QLgocHgzqktPy91U7tdyYKyn7k)
> 
> > I'm also pretty sure you can still block the falling rocks on DMT, though it has been a very long time since I tested all this
> 
> You can get a block to trigger but the breaking hitbox/a second boulder hits you anyway even if you hold the shield upwards, so it does not avoid damage the same way turtle shield would

All these should be fixed now. As for the DMT part, I think that is just to be expected if you want to use this then.

Pepper0ni commented:
This got pushed aside by the 9.0 push, conflicts need fixing
